### PR TITLE
Python: add buildPythonPackage.overridePythonPackage method.

### PR DESCRIPTION
**Pull request by FRidh**
This allows one to always override the call to `buildPythonPackage`.

The following example renamed `pandas` to `foo`:

```
with import <nixpkgs> {};

(let
python = let
    packageOverrides = self: super: {
      pandas = super.pandas.overridePythonPackage(old: {name="foo";});
    };
in pkgs.python35.override {inherit packageOverrides;};

in python.withPackages(ps: [ps.pandas])).env
```

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Closes https://github.com/NixOS/nixpkgs/issues/24154.

### CODE REVIEW

File: doc/languages-frameworks/python.md [hunk 1]
```diff
@@ -3,7 +3,7 @@
 ## User Guide
 
 Several versions of Python are available on Nix as well as a high amount of
-packages. The default interpreter is CPython 3.5.
+packages. The default interpreter is CPython 2.7.
```
Comment (Mic92) [inline doc/languages-frameworks/python.md]: I would back port this commit.

File: doc/languages-frameworks/python.md [hunk 2]
```diff
@@ -422,8 +422,8 @@ and in this case the `python35` interpreter is automatically used.
 
 ### Interpreters
 
-Versions 2.6, 2.7, 3.3, 3.4 and 3.5 of the CPython interpreter are available as respectively
-`python26`, `python27`, `python33`, `python34` and `python35`. The PyPy interpreter
+Versions 2.7, 3.3, 3.4, 3.5 and 3.6 of the CPython interpreter are available as
+respectively `python27`, `python33`, `python34`, `python35` and `python36`. The PyPy interpreter
```
Comment (Mic92) [inline doc/languages-frameworks/python.md]: I would back port this commit.

File: doc/languages-frameworks/python.md (hunk 1 above)
Comment (FRidh) [inline doc/languages-frameworks/python.md]: Absolutely!
Actually, I would like to get all of them in 17.03 :)

Comment (FRidh) [inline doc/languages-frameworks/python.md]: (Which is why I kept them together now)

File: doc/languages-frameworks/python.md [hunk 3]
```diff
@@ -546,6 +546,26 @@ All parameters from `mkDerivation` function are still supported.
...
+package to `foo`. We override first the Python interpreter and pass
+`packageOverrides` which contains the overrides for packages in the package set.
+
+```nix
+with import <nixpkgs> {};
+
+(let
+  python = let
+    packageOverrides = self: super: {
+      pandas = super.pandas.overridePythonPackage(old: {name="foo";});
```
Comment (Mic92) [inline doc/languages-frameworks/python.md]: A more practical example:

```
pandas = super.pandas.overridePythonPackage(old: rec {
  version = "0.19.1";
  src =  super.fetchPypi {
    pname = "pandas";
    inherit version;
    sha256 = "08blshqj9zj1wyjhhw3kl2vas75vhhicvv72flvf1z3jvapgw295";
  };
});
```

File: pkgs/top-level/python-packages.nix [hunk 4]
```diff
@@ -27,7 +27,27 @@ let
 
   mkPythonDerivation = makeOverridable( callPackage ../development/interpreters/python/mk-python-derivation.nix {
   });
-  buildPythonPackage = makeOverridable (callPackage ../development/interpreters/python/build-python-package.nix {
+
```
Comment (FRidh) [inline pkgs/top-level/python-packages.nix]: Note that this is basically a copy of `lib.makeOverrable`.

File: doc/languages-frameworks/python.md (hunk 3 above)
Comment (FRidh) [inline doc/languages-frameworks/python.md]: Hmm, I thought my example (and yours) was working but now it doesn't...
Did you try it?

Comment (FRidh) [inline doc/languages-frameworks/python.md]: I can use `.overridePythonPackage` inside say `python-packages.nix` but not in a separate `shell.nix`.

**Comment by FRidh** [comment]
This doesn't function. I'll have a look at it later.

**Pull request closed without merging**